### PR TITLE
Add Text/replace builtin

### LINKED ISSUE #1051: Having a way to apply string/text transformation like `replace`
As a usecase, I use dhall-kubernetes and I want to generate my config from a simple input. 

As an example, I want use the domain name `example.com` to populate my config.

But, I have a limitation due to kubernetes. I can't reuse the `.` (dot) char as `name` for the `namespace` type/object. Then, normally, with a configuration tool like `helm` I replace `.` with `-` to get a `namespace` named like `example-com`.

At the end, creating new strings by replacing char from a previous one is a go/no-go for me to use dhall (vs jsonnet). I'm not a pro in haskell even if I try. I will need assistance to make it work or community contribution. For sure, we can simply position this issue as a _feature request._ Thank you for any help ! 

PS : Sorry for my english.

### PR DESCRIPTION
**Pull request by alexhumphreys**
This attempts to standardise the `Text/replace` builtin.

There's a few parts I wasn't sure on (eg. the `B` case for the test `tests/parser/success/builtinsA.dhall`), and there's probably plenty other things I missed so best to look over this carefully.

Also, wasn't sure how best to describe this feature in the prose sections (doc strings in prelude, Language-tour.md, etc.). I feel the need to use the phrase "substring" but Dhall has `Text` not `String` so maybe that's misleading. So if anyone has suggestions to make those descriptions more clear I'd appreciate it.

Fixes #1051.

### CODE REVIEW

File: standard/beta-normalization.md [hunk 1]
```diff
@@ -661,13 +661,28 @@ Or in other words:
...
+`Text/replace` modifies a subset of a given `Text` literal. It takes 3 arguments, the `Text` literal subset to match, the `Text` literal replacement, and the `Text` literal in which to replace all matches:
+
+
+    f ⇥ Text/replace "bar" "baz" a ⇥ "foobaz123"
+    ────────────────────────────
+    f a ⇥ "foobar123"
+
+    f ⇥ Text/replace "-" "_" a ⇥ "foo_bar"
+    ────────────────────────────
+    f a ⇥ "foo-bar"
```
Comment (Gabriel439) [inline standard/beta-normalization.md:673]: I think the way I would do this is to define the behavior by induction:

```suggestion
    f ⇥ Text/replace needle replacement   a ⇥ ""
    ────────────────────────────────────────────
    f a ⇥ ""

    f ⇥ Text/replace "needle" replacement
    a ⇥ "needleHaystack…"
    Text/replace "needle" "Haystack…" → replacements…
    replacement ++ replacements… → e
    ─────────────────────────────────────────────────  ; Prefix matches
    f a ⇥ e

    f ⇥ Text/replace "needle" replacement
    a ⇥ "xHaystack…"
    Text/replace "needle" "Haystack…" → replacements…
    "x" ++ replacements… → e
    ─────────────────────────────────────────────────  ; Prefix does not match
    f a ⇥ e
```

Carefully note that this would permit replacing a substring with an abstract (interpolated expression).  If people object to that idea, then we can standardize a simpler version which doesn't support an `replacement` argument.  However, if we do support abstract replacement then we'd also want to add a test case equivalent to this assertion:

```dhall
λ(x : Text) → Text/replace "a" x "ab" ≡ "${x}b"
```

**Review by Gabriel439 — COMMENTED** [review]
Looks great so far!

I think one thing we need to decide is what to do if the substring to replace is `""`, in order to avoid an infinite loop.  Some possibilities I can think of are:

* Silently perform no replacement if the substring to replace is empty
* Return an `Optional` with a `None` result
* Require the substring to replace to be statically known at type-checking time and return a type error if empty

File: Prelude/Text/replace.dhall [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{-
+Replace a section of `Text` with another inside a `Text` literal.
+-}
+let replace
+    : ∀(needle : Text) → ∀(replacement : Text) → ∀(haystack : Text) → Text
+    = Text/replace
+
+let example0 = assert : replace "-" "_" "foo-bar" ≡ "foo_bar"
+
+let example1 = assert : replace "💣" "💥" "💣💣💣" ≡ "💥💥💥"
```
[the remark this answers is not in the document]
Comment (alexhumphreys) [reply Prelude/Text/replace.dhall:10]: Oooh good test, will add that. There's probably some other interesting tests to add (RTL character?), but not sure how exhaustive we need to be here.

**Comment by alexhumphreys** [comment]
As regards "", I'd vote for the first choice of do nothing. The second choice of returning an optional seems unergonomic. The third choice seems like an odd type error, given that "non-empty string" (or non-empty list for that matter) isn't a type that exists anywhere else in the language.

Just tried ruby to see what it does in this case and it was not what I was expecting:

```ruby
"foo".gsub "" "bar" == barfbarobarobar
```

... not sure we want to follow that example 😅 Though I can see a case to be made for `Text/replace "" "foo" "" === "foo"`.

I don't feel strongly about this though, whatever ye all prefer is cool.

File: standard/beta-normalization.md (hunk 1 above)
Comment (alexhumphreys) [reply standard/beta-normalization.md:673]: I'm not a pro at reading these beta normalisation rules so I will take me a while to comment on the suggested edit, but behaviour like `λ(x : Text) → Text/replace "a" x "ab" ≡ "${x}b"` is something I would want/expect from this feature.

I'm happy to add the suggested edit, but will wait a while to let others chime in on the issue.

File: Prelude/Text/replace.dhall (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexhumphreys) [reply Prelude/Text/replace.dhall:10]: Updated to add this test.

**Comment by alexhumphreys** [comment]
@Gabriel439 I've committed your suggested change defining normalisation in terms of recursion. I've also added a note to do nothing in the case of `""` as the substring to replace, and if we decide on some other behaviour I can update this later.

**Comment by Gabriel439** [comment]
@alexhumphreys: I'm taking vacation this week, but I'll review again afterwards

**Comment by alexhumphreys** [comment]
Ah nice, enjoy the vacation! 🌴

**Review by Nadrieril — APPROVED** [review]

File: docs/references/Built-in-types.md [hunk 3]
```diff
@@ -815,6 +815,23 @@ string:
...
+⊢ Text/replace "foo" "bar" "foobar"
+
+"barbar"
+```
+
+#### Type
+
+```
+──────────────────────────────────────
+Γ ⊢ Text/replace : Text → Text → Text → Text
```
Comment (Gabriel439) [inline docs/references/Built-in-types.md:832]: Naming the arguments for the inferred type using `∀`s might help people discern what each `Text` is supposed to be

```suggestion
Γ ⊢ Text/replace : ∀(needle : Text) → ∀(replacement : Text) → ∀(haystack : Text) → Text
```

File: docs/tutorials/Language-Tour.md [hunk 4]
```diff
@@ -910,6 +910,24 @@ You can concatenate `Text` literals using the `++` operator:
...
+a name where you need to replace dashes with underscores. That can be achieved
+with the following:
+
+```dhall
+⊢ Text/replace "-" "_" "foo-bar-baz"
+
+"foo_bar_baz"
+```
+
+You can also replace larger sections of text:
```
Comment (Gabriel439) [inline docs/tutorials/Language-Tour.md:923]: This is non-blocking, but I think a useful example to add here would be to show how to use `Text/replace` to strip newlines from a multi-line string, like this:

```dhall
Text/replace "\n" " " ''
  This now behaves like a YAML
  "folded" multi-line string literal where
  newlines are replaced with spaces
''
```

This idea came up in the FP Slack #dhall channel

File: standard/beta-normalization.md [hunk 5]
```diff
@@ -661,13 +661,45 @@ Or in other words:
...
+    "x" ++ replacements… → e
+    ─────────────────────────────────────────────────  ; Prefix does not match
+    f a ⇥ e
+
+If the substring to replace is empty (`""`), then no replacement is performed:
+
+
+    f ⇥ Text/replace "" replacement   a ⇥ "foo"
+    ────────────────────────────────────────────
+    f a ⇥ "foo"
```
Comment (Gabriel439) [inline standard/beta-normalization.md:690]: Could you move this judgement to be the first or second one?  The standard has an implicit "first judgement wins" convention that we've been using when more than one judgement applies, and we want this judgment to take precedence

**Review by Gabriel439 — COMMENTED** [review]
Could you also add a test for replacing an empty string?  That way, implementations won't forget to handle that case

Other than that, this looks to me like this proposal is basically ready to move out of Draft status

File: docs/tutorials/Language-Tour.md (hunk 4 above)
Comment (alexhumphreys) [reply docs/tutorials/Language-Tour.md:923]: Updated to add this.

File: standard/beta-normalization.md (hunk 5 above)
Comment (alexhumphreys) [reply standard/beta-normalization.md]: Updated.

**Comment by alexhumphreys** [comment]
@Gabriel439 I've added that empty string test. I still don't have the `B` case for `tests/parser/success/builtinsA.dhall` as I'm not sure what `Text/replace` gets parsed as, but aside from from my side it's also good to move this out of draft status.

**Comment by sjakobi** [comment]
Isn't the normalization still way under-specified? How are [composite characters](https://en.wikipedia.org/wiki/Unicode#Ready-made_versus_composite_characters) supposed to be handled, for example? (I believe @Profpatsch had some opinions about that in some discussion about `Text` operations)

**Comment by Gabriel439** [comment]
@sjakobi: I'd propose doing replacement at the grapheme level, as suggested by this test:

https://github.com/dhall-lang/dhall-lang/blob/595c8ca673966a1ab9b90f4a1cfd87789d660735/Prelude/Text/replace.dhall#L12

… and for correctly handling composite characters I'd suggest performing the replace if the "needle" and the prefix are both canonically equivalent, meaning that they have the same [normalization form](www.unicode.org/reports/tr15/), specifically Normalization Form C.

**Comment by sjakobi** [comment]
@Gabriel439 Sounds good to me!

**Comment by alexhumphreys** [comment]
@Gabriel439 @sjakobi I updated this with a note on using NFC. Well, I pretty much just took your comment from above. Does this need more detail or would that be clear enough?

File: Prelude/Text/replace.dhall [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+let example2 = assert : replace "👨" "👩" "👨‍👩‍👧‍👦" ≡ "👩‍👩‍👧‍👦"
```
Comment (Gabriel439) [inline Prelude/Text/replace.dhall:12]: Could you also add this as one of the test cases in the standard test suite?

**Review by Gabriel439 — APPROVED** [review]

File: Prelude/Text/replace.dhall (hunk 6 above)
Comment (alexhumphreys) [reply Prelude/Text/replace.dhall:12]: Yep, updated. Should I leave this in the prelude? Just wondering how easy this will be for all implementations, since most users importing the prelude won't need this particular replacement.

File: standard/beta-normalization.md [hunk 7]
```diff
@@ -661,13 +661,49 @@ Or in other words:
     f a ⇥ "\"…\\n…\\u0024…\\\\…\\\"…\\u0000…\""
 
 
-Otherwise, in isolation `Text/show` is in normal form:
+`Text/replace` modifies a subset of a given `Text` literal. It takes 3
```
Comment (sjakobi) [inline standard/beta-normalization.md:664]: Is the use of "subset" instead of "substring" here and below intentional?

File: standard/beta-normalization.md [hunk 8]
```diff
@@ -661,13 +661,49 @@ Or in other words:
...
+    a ⇥ "xHaystack…"
+    Text/replace "needle" "Haystack…" → replacements…
+    "x" ++ replacements… → e
+    ─────────────────────────────────────────────────  ; Prefix does not match
+    f a ⇥ e
+
+
+A given "needle" string matches a given subset of the "haystack" string if they
+are canonically equivalent, meaning that they have the same normalization form,
+specifically [Normalization Form C](https://unicode.org/reports/tr15/#Norm_Forms).
```
Comment (sjakobi) [inline standard/beta-normalization.md:695]: It would be good to have tests for the normalization semantics – e.g. "failing"/ineffective tests that would require Compatibility Decomposition, but also tests that require Canonical Decomposition and Canonical Composition. Maybe we can simply use some of the examples from the unicode.org page.

File: Prelude/Text/replace.dhall (hunk 6 above)
Comment (Gabriel439) [reply Prelude/Text/replace.dhall:12]: I think it is also worth keeping in the Prelude since this is useful documentation for the end user.  These examples will also show up in the generated docs

File: standard/beta-normalization.md (hunk 8 above)
Comment (alexhumphreys) [reply standard/beta-normalization.md:695]: Hmmm I'm not sure what tests would cover what you're saying. 

(The Unicode page is mostly pictures of characters I can't type, so this is probably about to get super confusing 😅)

From the [unicode](https://unicode.org/reports/tr15/#Norm_Forms) docs, there's say the example with "d" and "q" with accents (the circles above and below). It seems "d with dots above and below" gets NFC normalised to "d with dot below"+"dot above". Whereas "q with dots above and below" gets NFC normalised to "q"+"dot below"+"dot above".

So are you saying to make a test out of that? Like `replace "q" "d" "<q with dots above and below>"` which would be... even looking at the docs I'm not sure what would happen in this case, probably a good sign it should be a test 😅

And what would a failing test be? Something that throws like a type error, or something that fails to replace? Both?

Comment (Gabriel439) [reply standard/beta-normalization.md:695]: @alexhumphreys: I think the test would be more like:

```dhall
Text/replace "a\u{0301}" "b" "\u{00e1}c" = "bc"
```

… where "a\u{0301}" and "\u{00e1}" both represent `á` and have the same normal form.

However, I think the way I would approach the problem is to split this into two separate fixes:

* Define the rule for prefix matching in terms of the equivalence rule

* Change how `Text` literals are binary encoded so that they are normalized before encoding

  … which will transitively fix the equivalence rule to treat NFC-equivalent `Text` literals as the same
  
In other words, you can define the judgment for `Text/replace` so that the replacement occurs if and only if the "needle" and the "prefix" are equivalent according to `≡`, like this:

```
    f ⇥ Text/replace "needle" replacement
    a ⇥ "prefixHaystack…"
    "needle" ≡ "prefix"
    Text/replace "needle" "Haystack…" → replacements…
    replacement ++ replacements… → e
    ─────────────────────────────────────────────────  ; Prefix matches
    f a ⇥ e 
```

… and then whether or not the prefix matches defers to the rules for `Text` equivalence.

Then you update [the binary encoding of `Text` literals](https://github.com/dhall-lang/dhall-lang/blob/master/standard/binary.md#text) to say that they are NFC normalized before encoding, like this:

```
normalization-form-C("a") = a
encode(b₀) = b₁
normalization-form-C("c") = c
encode(d₀) = d₁
normalization-form-C("e") = e
…
normalization-form-C("x") = x
encode(y₀) = y₁
normalization-form-C("z") = z
───────────────────────────────────────────────────────────────────────────────────
encode("a${b₀}c${d}e…x${y₀}z") = [ 18, a, b₁, c, d₁, e, …, x, y₁, z ]
```

… and then if you do that then `Text/replace` will do the "right thing".

Doing it that way would have the bonus of also fixing other corner cases of the language.  For example, currently the Haskell implementation treats the two representations of `á` as not equivalent:

```dhall
⊢ assert : "\u00e1" === "a\u0301"

Error: Assertion failed

"áá"

1│ assert : "\u00e1" === "a\u0301"

(input):1:1
```

… but if you change `Text` literals to be normalized before encoding then it would treat them the same.

**Comment by MonoidMusician** [comment]
I'm wondering if NFD would actually be a better choice? That way you can replace accents and other combining characters, e.g. `Text/replace "\u{0301}" "\u{0304}" "á" = "ā"`. NFD will certainly provide more options for performing replacements, which would be what I expect. But I'm also a little wary of normalizing user inputs by default …

**Comment by Gabriel439** [comment]
@MonoidMusician: Yeah, that's a good point.  Normalization form D is probably a better choice

**Comment by Gabriel439** [comment]
@alexhumphreys: I have a PR open with all of my suggested changes in: https://github.com/alexhumphreys/dhall-lang/pull/1

I decided to skip handling normalization for now (and adding a test making that explicit).  I can standardize support for Unicode normalization in a separate proposal, to keep this pull request simpler.

**Comment by Gabriel439** [comment]
I also have a matching change to the Haskell implementation up here: https://github.com/dhall-lang/dhall-haskell/pull/2063

**Comment by alexhumphreys** [comment]
@Gabriel439 thanks for the PR, those beta normalisation rules are a lot clearer. It also thought me a bit about string interpolation: didn't know it would get normalised under a lambda binding before the lambda is applied.

Splitting the Unicode handling to a different PR is cool too. Would `let example2 = assert : replace "👨" "👩" "👨‍👩‍👧‍👦" ≡ "👩‍👩‍👧‍👦"` still remain in the prelude for this PR? I _think_ the output of that wouldn't change between NFC and NFD normalisation, so it should be ok.

File: standard/beta-normalization.md (hunk 7 above)
Comment (alexhumphreys) [reply standard/beta-normalization.md]: Nope, updated.

**Comment by Gabriel439** [comment]
@alexhumphreys: I think that example would be the same regardless of how we handle normalization

**Review by Gabriel439 — APPROVED** [review]

File: docs/references/Built-in-types.md [hunk 9]
```diff
@@ -815,6 +815,23 @@ string:
 "\"\\n🎉\""
 ```
 
+### Function `Text/replace`
+
+The `Text/replace` built-in function modifies a subset of a given `Text` literal. It takes 3 arguments, the `Text` literal subset to match, the `Text` literal replacement, and the `Text` literal in which to replace all matches:
```
Comment (sjakobi) [inline docs/references/Built-in-types.md:820]:
```suggestion
The `Text/replace` built-in function modifies a substring of a given `Text` literal. It takes 3 arguments, the `Text` literal substring to match, the `Text` literal replacement, and the `Text` literal in which to replace all matches:
```

**Review by sjakobi — APPROVED** [review]
LGTM apart from one more wibble. Thanks! :)

File: docs/references/Built-in-types.md (hunk 9 above)
Comment (alexhumphreys) [reply docs/references/Built-in-types.md]: Good catch! Updated.

**Comment by Gabriel439** [comment]
@alexhumphreys: You should be clear to merge this

**Comment by alexhumphreys** [comment]
@Gabriel439 sweet! Any 3 day waiting period or can I just hit the button?

**Comment by Gabriel439** [comment]
@alexhumphreys: Since you don't have a majority of approvals but you also don't have a majority of disapprovals then the policy is 7 days since the PR was ready (i.e. not a draft PR) and 1 day since last code change.  For more details, see: https://github.com/dhall-lang/dhall-lang/blob/master/.github/CONTRIBUTING.md#how-do-changes-get-approved

**Comment by alexhumphreys** [comment]
Ah cool, so it should be good to be merged. Looks like I don't have write access so maybe one of ye wants to hit the button? Also there's 25 commits in this PR now, do ye have a policy of squashing commits or should I go in and tidy up the history a bit?

**Comment by philandstuff** [comment]
This repo enforces squash-on-merge. We should get you the commit bit so you can merge yourself 👍🏻

**Comment by philandstuff** [comment]
@alexhumphreys I’ve granted you wire access now

**Comment by alexhumphreys** [comment]
Sweet! Thanks @philandstuff, and everyone else too for helping on this 🙂

**Comment by Gabriel439** [comment]
@alexhumphreys: Thank you, too!  Also, now that this is merged you can submit an invoice here for $200 and one of us will approve it: https://opencollective.com/dhall/expenses/new

**Comment by alexhumphreys** [comment]
Thanks, but it's ok! I've been meaning to donate to that Dhall fund so consider this a contribution 🙂

File: tests/normalization/success/unit/TextReplaceAbstractA.dhall [hunk 10]
```diff
@@ -0,0 +1,4 @@
+{- This test verifies that an implementation correctly permits both the
+   "replacement" and the "haystack" to be abstract.
+-}
+λ(x : Text) → λ(y : Text) → Text/replace "a" "-${x}-" "_a_${y}_a_"
```
Comment (basile-henry) [inline tests/normalization/success/unit/TextReplaceAbstractA.dhall:4]: I don't really understand how the haystack can be abstract. Don't we want to replace occurrences of the needle in the abstract section too? i.e. What happens if `y` is the string `"a"`?

**Review by basile-henry — COMMENTED** [review]
Sorry, I'm a bit late for a review. I just have a question as I am attempting to implement this in `dhall-rust`. :sweat_smile:

File: tests/normalization/success/unit/TextReplaceAbstractA.dhall (hunk 10 above)
Comment (Gabriel439) [reply tests/normalization/success/unit/TextReplaceAbstractA.dhall:4]: @basile-henry: Oh, I see what you mean.  Allowing the haystack to be abstract means that β-reduction is no longer confluent.  For right now, just ignore that test and I can put up a change to require the haystack to be non-abstract.
